hoptimator-avro/src/main/java/com/linkedin/hoptimator/avro/AvroConverter.java

```diff
@@ -0,0 +1,149 @@
+package com.linkedin.hoptimator.avro;
+
+import org.apache.avro.Schema;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeImpl;
+import org.apache.calcite.rel.type.RelDataTypeSystem;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.rel.type.RelProtoDataType;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.type.SqlTypeFactoryImpl;
+
+import java.util.AbstractMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/** Converts between Avro and Calcite's RelDataType */
+public final class AvroConverter {
+
+  private AvroConverter() {
+  }
+
+  public static Schema avro(String namespace, String name, RelDataType dataType) {
+    if (dataType.isStruct()) {
+      List<Schema.Field> fields = dataType.getFieldList().stream()
+        .map(x -> new Schema.Field(sanitize(x.getName()), avro(namespace, x.getName(), x.getType()), describe(x), null))
+        .collect(Collectors.toList());
+      return createAvroSchemaWithNullability(Schema.createRecord(sanitize(name), dataType.toString(), namespace, false, fields),
+        dataType.isNullable());
+    } else {
+      switch (dataType.getSqlTypeName()) {
+      case INTEGER:
+        return createAvroTypeWithNullability(Schema.Type.INT, dataType.isNullable());
+      case SMALLINT:
+        return createAvroTypeWithNullability(Schema.Type.INT, dataType.isNullable());
+      case BIGINT:
+        return createAvroTypeWithNullability(Schema.Type.LONG, dataType.isNullable());
+      case VARCHAR:
+        return createAvroTypeWithNullability(Schema.Type.STRING, dataType.isNullable());
+      case FLOAT:
+        return createAvroTypeWithNullability(Schema.Type.FLOAT, dataType.isNullable());
+      case DOUBLE:
+        return createAvroTypeWithNullability(Schema.Type.DOUBLE, dataType.isNullable());
+      case CHAR:
+        return createAvroTypeWithNullability(Schema.Type.STRING, dataType.isNullable());
+      case BOOLEAN:
+        return createAvroTypeWithNullability(Schema.Type.BOOLEAN, dataType.isNullable());
+      case ARRAY:
+        return createAvroSchemaWithNullability(Schema.createArray(avro(null, null, dataType.getComponentType())),
+          dataType.isNullable());
+  // TODO support map types
+  // Appears to require a Calcite version bump
+  //    case MAP:
+  //      return createAvroSchemaWithNullability(Schema.createMap(avroPrimitive(dataType.getValueType())), dataType.isNullable());
```

Comment:
I think I held back the calcite upgrade so that the calcite version would match the version Flink is using. when we upgrade Flink, we should be able to upgrade calcite and uncomment this. It doesn't really matter whether they're using the same version or not, I just wanted to avoid having two versions of the same dependency.